src/metabase/api/user_key_value.clj

```diff
@@ -0,0 +1,59 @@
+(ns metabase.api.user-key-value
+  (:require [compojure.core :refer [GET PUT DELETE]]
+            [malli.core :as mc]
+            [malli.experimental.time.transform :as mett]
+            [malli.transform :as mtx]
+            [metabase.api.common :as api]
+            [metabase.models.user-key-value :as user-key-value]
+            [metabase.models.user-key-value.types :as types]
+            [metabase.util.malli.schema :as ms]))
+
+(api/defendpoint PUT "/namespace/:namespace/key/:key"
+  "Upsert a KV-pair for the user"
+  [:as {{v :value
+         expires_at :expires_at} :body
+        {namespace :namespace
+         k :key} :params}]
+  {k ms/NonBlankString
+   v :any
+   namespace ms/NonBlankString
+   expires_at [:maybe :metabase.lib.schema.literal/string.datetime]}
+  (try (user-key-value/put! api/*current-user-id* (mc/coerce ::types/user-key-value
+                                                             {:key k
+                                                              :namespace namespace
+                                                              :value v
+                                                              :expires-at expires_at}
+                                                             (mtx/transformer
+                                                              (mtx/default-value-transformer)
+                                                              (mett/time-transformer)
+                                                              {:name :api-request})))
+       (catch Exception e
+         (when (= (:type (ex-data e))
+                  ::mc/coercion)
+           (api/check-400 false))
+         (throw e))))
+
+(api/defendpoint GET "/namespace/:namespace/key/:key"
+  "Get a value for the user"
+  [namespace key]
+  {key ms/NonBlankString
+   namespace ms/NonBlankString}
+  (user-key-value/retrieve api/*current-user-id* namespace key))
+
+(api/defendpoint GET "/namespace/:namespace"
+  "Get the value for multiple keys in a namespace"
+  [:as {{:strs [keyyy]}
+        :query-params :as prms}
+   namespace]
+  {keyyy [:maybe (ms/QueryVectorOf ms/PositiveInt)]
```

Comment:
This is broken - and if I'm not mistaken, it seems like it's actually broken everywhere that we try to do this? For example, in search or collection items endpoints, it seems like we actually only get one item from the query params, rather than all of them. 😬 

To get this merged, we could skip the bulk endpoint for now, or maybe just return all the keys in the namespace?

Reply:
I just wrote a test to check it — I'm pretty sure I've checked it manually, and it passes vectors even for a single argument (see `serialization/api.clj`) — but a test won't harm; and it seems to be working, see #51385

So it's working in at least one other case...